File: python/phone-number/phone_number.py
import re


class PhoneNumber:
    def __init__(self, number: str):
        number = re.sub(r"\D", "", number)

        if len(number) < 10 or len(number) > 11:
            raise ValueError("Invalid phone number")

        elif len(number) == 11:
            if number[0] != "1":
                raise ValueError("Invalid phone number")

            number = number[1:]

        if number[0] < "2" or number[3] < "2":
            raise ValueError("Invalid phone number")

        self._number: str = number

    @property
    def number(self) -> str:
        return self._number

    @property
    def area_code(self) -> str:
        return self.number[:3]

    def pretty(self) -> str:
        return f"({self.number[:3]}) {self.number[3:6]}-{self.number[6:]}"

File: cpp/phone-number/phone_number.cpp
#include "phone_number.h"

#include <regex>
#include <exception>

using namespace std;

namespace phone_number
{
    phone_number::phone_number(const string &number)
    {
        regex re(R"(\D+)");
        _number = regex_replace(number, re, "");

        if (_number.size() < 10 || _number.size() > 11)
        {
            throw domain_error("Invalid phone number");
        }

        if (_number.size() == 11)
        {
            (_number[0] == '1') ? _number.erase(0, 1) : throw domain_error("Invalid phone number");
        }

        if (_number[0] < '2' || _number[3] < '2')
        {
            throw domain_error("Invalid phone number");
        }
    }

    string phone_number::number() const
    {
        return _number;
    }

    string phone_number::area_code() const
    {
        return _number.substr(0, 3);
    }

    phone_number::operator std::string() const
    {
        return "(" + _number.substr(0, 3) + ") " + _number.substr(3, 3) + "-" + _number.substr(6, 4);
    }
} // namespace phone_number
